### ISSUE TITLE
F::mse_loss accepts reduction as &str — should use an enum for type safety

### ISSUE THREAD

**Issue by IAmMahiro** [cool-japan/torsh#45]
Hi! Small API ergonomics note.

## Current API

```rust
let loss = torsh::F::mse_loss(&fwd, &y, "mean")?;
```

The reduction mode is passed as a string literal. This means:
- Typos like "meen" or "avg" compile fine and fail at runtime (or silently default)
- No IDE autocompletion for valid options
- Not idiomatic Rust — enums exist precisely for closed sets of options

## Suggestion

```rust
pub enum Reduction {
    Mean,
    Sum,
    None,
}

// Usage:
let loss = torsh::F::mse_loss(&fwd, &y, Reduction::Mean)?;
```

If backward compatibility is a concern, From<&str> could be implemented with a deprecation warning:

```rust
impl TryFrom<&str> for Reduction {
    type Error = AutogradError;
    fn try_from(s: &str) -> Result<Self> { ... }
}

#[deprecated(note = "Use Reduction::Mean instead of \"mean\"")]
pub fn mse_loss_str(a: &Tensor, b: &Tensor, reduction: &str) -> Result<Tensor> { ... }
```

---
This likely applies to other loss functions too (cross_entropy, l1_loss, etc.). Happy to audit and list them if helpful.
Thanks!

**Comment by IAmMahiro** [comment]
Aha, found that torsh::F has 

```rust
pub struct MSELoss {
    pub reduction: Reduction,
}
```

so it shouldn't be hard to reuse the same Reduction enum for functional losses.

**Comment by IAmMahiro** [comment]
> Aha, found that torsh::F has
> 
> pub struct MSELoss {
>     pub reduction: Reduction,
> }
> 
> so it shouldn't be hard to reuse the same Reduction enum for functional losses.

Like that:

```rust
use crate::functional::Reduction;

fn apply_reduction(
    loss: &Tensor,
    reduction: &Reduction,
    ignore_index: Option<i64>,
    target_data: &[i64],
) -> Result<Tensor> {
    match reduction {
        &Reduction::Mean => {
            if let Some(ignore_idx) = ignore_index {
                // Count non-ignored samples
                let valid_count =
                    target_data.iter().filter(|&&idx| idx != ignore_idx).count() as f32;
                if valid_count > 0.0 {
                    let sum = loss.sum()?;
                    let count_tensor = torsh_tensor::creation::full(&[1], valid_count)?;
                    sum.div(&count_tensor)
                } else {
                    loss.mean(None, false)
                }
            } else {
                loss.mean(None, false)
            }
        }
        &Reduction::Sum => loss.sum(),
        &Reduction::None => Ok(loss.clone()),
    }
}
```

**Comment by cool-japan** [comment]
Thanks @IAmMahiro — the stringly-typed reduction is real, and the inconsistency is worse than the report assumes.

**One correction first.** Typos do not silently default. Every reduction dispatch site in the workspace returns `Err` on an unknown string — all 14 of them, e.g. `crates/torsh-nn/src/functional/loss.rs:1069`, `:490`, `:712`, `:779`, `:879`, `crates/torsh-functional/src/regularization.rs:160`, `crates/torsh-functional/src/loss/classification.rs:148`. So `"meen"` is a clean runtime error, not a silently wrong loss. This is an ergonomics and consistency problem, not a correctness one.

**What is actually wrong is bigger than one signature.** A `Reduction` enum already exists — three of them, in fact:
- `Reduction { None, Mean, Sum }` — `crates/torsh-nn/src/functional/loss_advanced.rs:14`, with a fallible `from_str` at `:25`
- `ReductionType { None, Mean, Sum }` — `crates/torsh-functional/src/loss/common.rs:10`
- `Reduction { None, Mean, Sum, BatchMean }` — `crates/torsh-vision/src/ops/analysis/mod.rs:20`

Meanwhile 32 public functions still take `reduction: &str`, and ten loss names exist in both spellings across sibling crates — `mse_loss`, `l1_loss`, `smooth_l1_loss`, `kl_div`, `focal_loss`, `multi_margin_loss`, `contrastive_loss`, `cosine_embedding_loss`, `triplet_margin_loss`, `binary_cross_entropy{,_with_logits}` — string-typed in torsh-nn, enum-typed in torsh-functional. Both land in one namespace, because `crates/torsh/src/lib.rs:549-560` glob-imports both into `torsh::F`. The two READMEs even document the same call with different types:

- `crates/torsh-nn/README.md:92` — `F::mse_loss(&predictions, &targets, "mean")?`
- `crates/torsh-functional/README.md:86` — `F::mse_loss(&predictions, &targets, ReductionType::Mean)?`

So the accurate framing is unification, not introduction.

**Planned direction:**
- One canonical `Reduction { None, Mean, Sum, BatchMean }` in **torsh-core**. torsh-nn and torsh-functional are siblings with no edge between them, so the shared ancestor is the only legal home for a single type.
- `impl FromStr` returning `Result`, matching the 8 existing `FromStr` impls in the workspace (there are zero `TryFrom<&str>` impls, so `FromStr` is the house pattern here rather than the `TryFrom` you sketched).
- Re-alias `torsh_nn::functional::Reduction` and `torsh_functional::ReductionType` to it so both names keep compiling, then migrate the 32 signatures.
- PyO3 entry points (`crates/torsh-ffi/src/functional.rs:163,241,295`, `crates/torsh-python/src/functional.rs:155,172,200`) stay stringly-typed and parse at the boundary — `crates/torsh-python/src/functional.rs:227` already does this correctly.

**One blocker worth naming**, since it would bite a naive migration: `kl_div` (`crates/torsh-nn/src/functional/loss.rs:452`) accepts `"batchmean"` (arm at `:497`), which neither candidate enum has. Swapping to an enum without adding `BatchMean` first would silently drop a supported, PyTorch-compatible mode.

And a note on the back-compat sketch: an `impl Into<Reduction>` parameter would be the wrong shape here, because `From<&str>` is infallible and would force a panic or a silent default on `"meen"` — manufacturing exactly the bug this issue suspected. The fallible `FromStr` keeps that honest.

Yes to your offer to audit the other losses — though the inventory is done: 20 in `crates/torsh-nn/src/functional/loss.rs`, 3 in `crates/torsh-functional/src/loss/classification.rs`, 5 in `crates/torsh-functional/src/regularization.rs`, 1 in `utils.rs`, 3 `*_configured` wrappers in `crates/torsh-nn/src/functional.rs`. If you want to help, the higher-value contribution would be the `torsh-core` enum plus the `BatchMean` handling.

Tracked in `TODO.md` under "GitHub Issue Triage Findings" and targeted for the **0.2.1** release; the signature change is accepted as breaking. Not implemented yet — I will follow up here when it lands.

**Comment by IAmMahiro** [comment]
Thanks for the detailed breakdown. I'd be happy to take this on.
I'll work on adding a unified ```Reduction``` enum to ```torsh-core``` (with ```BatchMean```).
Will try to implement ```FromStr``` following the existing workspace pattern.
Will add re-exports for compatibility.
And I'll work on migrating all 32 functions from ```&str``` to ```Reduction```.

**Comment by IAmMahiro** [comment]
Heya, I finished adding a unified Reduction. Please check my pull request.
